src/styles.rs

```diff
@@ -0,0 +1,536 @@
+use std::collections::HashMap;
+use std::fs;
+
+pub mod converter;
+pub mod css_colors;
+mod property_list;
+mod shorthands;
+
+/// A declarationProperty defines a single value for a property (color: red;). It consists of the value, origin, importance, location and
+/// specificity of the declaration.
+#[derive(Debug)]
+pub struct DeclarationProperty {
+    /// The actual value of the property
+    pub value: String,
+    /// Origin of the declaration (user stylesheet, author stylesheet etc)
+    pub origin: CssOrigin,
+    /// Whether the declaration is !important
+    pub important: bool,
+    /// The location of the declaration in the stylesheet (name.css:123) or empty
+    pub location: String,
+    /// The specificity of the selector that declared this property
+    pub specificity: Specificity,
+}
+
+impl DeclarationProperty {
+    /// Priority of the declaration based on the origin and importance as defined in https://developer.mozilla.org/en-US/docs/Web/CSS/Cascade
+    fn priority(&self) -> u8 {
+        match self.origin {
+            CssOrigin::UserAgent => {
+                if self.important {
+                    7
+                } else {
+                    1
+                }
+            }
+            CssOrigin::User => {
+                if self.important {
+                    6
+                } else {
+                    2
+                }
+            }
+            CssOrigin::Author => {
+                if self.important {
+                    5
+                } else {
+                    3
+                }
+            }
+        }
+    }
+}
+
+impl PartialEq<Self> for DeclarationProperty {
+    fn eq(&self, other: &Self) -> bool {
+        self.priority() == other.priority()
+    }
+}
+
+impl PartialOrd<Self> for DeclarationProperty {
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+        Some(self.cmp(other))
+    }
+}
+
+impl Eq for DeclarationProperty {}
+
+impl Ord for DeclarationProperty {
+    fn cmp(&self, other: &Self) -> Ordering {
+        self.priority().cmp(&other.priority())
+    }
+}
+
+///
+#[derive(Debug)]
+pub struct ValueEntry {
+    /// List of all declared values for this property
+    pub declared: Vec<DeclarationProperty>,
+    /// Cascaded value
+    pub cascaded: Option<String>,
+    pub specified: String,
+    pub computed: String,
+    pub used: String,
+    pub actual: String,
+}
+
+impl Default for ValueEntry {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl ValueEntry {
+    pub fn new() -> Self {
+        Self {
+            declared: Vec::new(),
+            cascaded: None,
+            specified: "".into(),
+            computed: "".into(),
+            used: "".into(),
+            actual: "".into(),
+        }
+    }
+}
+
+/// Map of all declared values for a single node
+pub type CssMapEntry = HashMap<String, ValueEntry>;
+
+/// Map of all declared values for all nodes in the document
+pub type CssMap = HashMap<NodeId, CssMapEntry>;
```

Comment:
I guess, this can be optimized when we just use one HashMap and basically do this
```suggestion
/// Map of all declared values for all nodes in the document
pub type CssMap = HashMap<(NodeId, String), ValueEntry>;
```

Reply:
Okay, then it's probably not quite possible. I'm actually not sure, if a HashMap is best here. Maybe we can use a custom data structure for that